extensions/csv/src/main/java/com/hazelcast/jet/csv/impl/CsvReadFileFnProvider.java

```diff
@@ -47,18 +53,49 @@
 )
 public class CsvReadFileFnProvider implements ReadFileFnProvider {
 
+    @SuppressWarnings("unchecked")
     @Nonnull
     @Override
     public <T> FunctionEx<Path, Stream<T>> createReadFileFn(@Nonnull FileFormat<T> format) {
         CsvFileFormat<T> csvFileFormat = (CsvFileFormat<T>) format;
         Class<?> formatClazz = csvFileFormat.clazz(); // Format is not Serializable
 
         return path -> {
-            ObjectReader reader = new CsvMapper().readerFor(formatClazz != null ? formatClazz : Map.class)
-                                                 .withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
-                                                 .with(CsvSchema.emptySchema().withHeader());
             FileInputStream fis = new FileInputStream(path.toFile());
-            return StreamSupport.<T>stream(Spliterators.spliteratorUnknownSize(reader.readValues(fis), ORDERED), false)
+            MappingIterator<T> iterator;
+            Function<T, T> projection = r -> r;
+            if (formatClazz == String[].class) {
+                ObjectReader reader = new CsvMapper().enable(Feature.WRAP_AS_ARRAY)
+                                                     .readerFor(String[].class)
+                                                     .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
+                iterator = reader.readValues(fis);
+                if (!iterator.hasNext()) {
+                    throw new JetException("Header row missing in " + path);
+                }
+                String[] header = (String[]) iterator.next();
+                List<String> fieldList = csvFileFormat.stringArrayFieldList();
+                if (fieldList != null) {
+                    int[] simpleFieldMap = createSimpleFieldMap(fieldList, header);
```

Comment:
I think you could avoid projection if _header_ == _stringArrayFieldList_.

Reply:
fixed